File: src/generator/tile.rb
class Tile
  attr_reader :kind, :items

  def limited
    [:B, :E, :V, :M]
  end

  def initialize(kind)
    @kind = kind.to_sym
    @items = []
  end

  def can_place?(item)
    item = item.to_sym

    # Only grass tiles can contain items
    return false if kind == :f

    # If there are no items, then it's okay
    return true if items.empty?

    # Only one item of each kind per terrain
    return false if items.include?(item)

    # If item is limited && we already have one limited
    return false if limited.include?(item) && !(items & limited).empty?

    true
  end

  def place(item)
    item = item.to_sym

    unless can_place?(item)
      raise StandardError.new("Can't place the item #{item} on this tile #{kind} [#{items.join(',')}]")
    end

    items.push(item)

    self
  end

  def to_s
    kind
  end
end

File: src/logic/tile.py
class Tile:
  def __init__(self, kind):
    self.kind = kind
    self.items = []

  def add_item(self, item):
    self.items.append(item)

  def remove_item(self, item):
  	self.items = [i for i in self.items if i != item]

  def is_safe(self):
    return not self.is_danger()

  def is_danger(self):
    return 'B' in self.items or 'E' in self.items or 'V' in self.items

  def is_deadly(self):
    return 'B' in self.items or 'E' in self.items
